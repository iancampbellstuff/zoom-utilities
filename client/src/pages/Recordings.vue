<!-- See https://codepen.io/b0otable/pen/PozWLYR for reference -->
<template>
    <q-page id="recordings" class="row items-start justify-evenly">
        <q-form id="form" class="q-gutter-lg">
            <q-toolbar-title class="text-center">Recordings</q-toolbar-title>
            <q-table
                :columns="columns"
                :filter-method="(customFilter as any)"
                :filter="filter"
                :rows="data"
                outlined
                row-key="id"
                standout
                title="Recordings"
                v-model:pagination="paginationOptions"
            >
                <template v-slot:top>
                    <div style="width: 100%" class="row">
                        <div :class="!!currentUserId ? 'col-3' : 'col-5'">
                            <q-select
                                :options="userIds"
                                @update:model-value="onChangeUserId"
                                fill-input
                                label="Select an Account:"
                                v-model="currentUserId"
                            />
                        </div>
                        <q-btn
                            @click="onRefresh"
                            color="secondary"
                            :disable="!currentUserId"
                            flat
                            icon="refresh"
                            label="Refresh"
                            no-caps
                            wait-for-ripple
                        />
                    </div>
                    <div style="width: 100%" class="row">
                        <div class="col-12">
                            <q-input
                                color="primary"
                                debounce="100"
                                v-model="search"
                                @keyup="onSearch"
                            >
                                <template v-slot:append>
                                    <q-icon name="search" />
                                </template>
                            </q-input>
                        </div>
                    </div>
                </template>
                <template v-slot:body="props">
                    <q-tr :props="props">
                        <q-td key="topic" :title="props.row.topic">{{
                            truncate(props.row.topic)
                        }}</q-td>
                        <q-td
                            key="recording_start"
                            :title="getFormattedDate(props.row.recording_start)"
                            >{{
                                truncate(
                                    getFormattedDate(
                                        props.row.recording_start
                                    ).toString()
                                )
                            }}</q-td
                        >
                        <q-td key="play_url">
                            <a
                                href="javascript:void(0);"
                                @click="
                                    () => {
                                        play = true;
                                        playUrl = props.row.play_url;
                                    }
                                "
                            >
                                {{ truncate(props.row.play_url) }}
                            </a>
                            <q-dialog v-model="play">
                                <q-card style="min-width: 350px">
                                    <q-card-section>
                                        <div class="text-h6">
                                            Copy or open this URL?
                                        </div>
                                    </q-card-section>
                                    <q-card-section
                                        class="q-pt-none bg-grey-4 scroll"
                                    >
                                        <q-input
                                            borderless
                                            readonly
                                            type="textarea"
                                            v-model="playUrl"
                                        ></q-input>
                                    </q-card-section>
                                    <q-card-actions
                                        class="text-primary justify-end"
                                    >
                                        <q-btn
                                            @click="onCopy"
                                            flat
                                            label="Copy"
                                            v-close-popup
                                        ></q-btn>
                                        <q-btn
                                            @click="onOpen"
                                            flat
                                            label="Open"
                                            v-close-popup
                                        ></q-btn>
                                        <q-btn
                                            flat
                                            label="Cancel"
                                            v-close-popup
                                        ></q-btn>
                                    </q-card-actions>
                                </q-card>
                            </q-dialog>
                        </q-td>
                        <q-td key="download_url">
                            <a :href="props.row.download_url" target="_blank">
                                {{ truncate(props.row.download_url) }}
                            </a>
                        </q-td>
                    </q-tr>
                </template>
            </q-table>
        </q-form>
    </q-page>
</template>

<style scoped lang="scss">
#recordings {
    #form {
        .q-toolbar__title {
            padding: 15px;
        }
        .q-table__container {
            width: 100%;
            .q-table__top {
                label {
                    overflow: hidden;
                }
            }
        }
    }
}
</style>

<script setup lang="ts">
// externals
import { computed, onBeforeMount, onMounted, ref } from 'vue';
import { useQuasar } from 'quasar';
// stores
import { useRecordingsStore } from '../stores';
// types
import { IZoomMeetingRecordingsResponseItem } from '@zoom-utilities/common/src/types';
// utils
import {
    copyInput,
    ELogLevel,
    getRecordings,
    getUserIds,
    toast
} from '../utils';
import { getFormattedDate, truncate } from '@zoom-utilities/common/src/utils';

interface IPaginationOptions {
    rowsPerPage: number;
}

const $q = useQuasar();
const paginationOptions = ref<IPaginationOptions>({
    rowsPerPage: 10
});
const columns: any = [
    {
        align: 'left',
        field: 'topic',
        label: 'Meeting Name',
        name: 'topic',
        required: true,
        sort: (a: string, b: string) => a.localeCompare(b),
        sortable: true
    },
    {
        align: 'left',
        field: 'recording_start',
        label: 'Recording Date',
        name: 'recording_start',
        required: true,
        sort: (a: Date, b: Date) => b.getTime() - a.getTime(),
        sortable: true
    },
    {
        align: 'left',
        field: 'play_url',
        label: 'Play URL',
        name: 'play_url',
        required: true
    },
    {
        align: 'left',
        field: 'download_url',
        label: 'Download URL',
        name: 'download_url',
        required: true
    }
];
const store = useRecordingsStore();
const currentUserId = ref('');
const data = ref<IZoomMeetingRecordingsResponseItem[]>([]);
const errorOccurred = ref(false);
const search = ref('');
const userIds = ref<string[]>([]);
const play = ref(false);
const playUrl = ref('');

onBeforeMount(() => {
    userIds.value = store.userIds;
    if (!userIds.value?.length) {
        errorOccurred.value = false;
        $q.loading.show();
        getUserIds()
            .then((ids: string[]) => {
                userIds.value = ids;
                store.setUserIds(userIds.value);
            })
            .catch(() => {
                errorOccurred.value = true;
                showErrorMessage();
            })
            .finally(() => {
                $q.loading.hide();
            });
    }
});

onMounted(() => {
    currentUserId.value = store.currentUserId;
    data.value = store.recordings;
    search.value = store.search;
});
const filter = computed(() => {
    return search.value;
});

const customFilter = (rows: any[], searchValue: string) => {
    const filteredRows = rows.filter(
        (row: { [s: string]: unknown } | ArrayLike<unknown>) => {
            let resultFound = true;
            if (searchValue !== '') {
                const rowValues = Object.entries(row).reduce(
                    (values: string[], [key, value]) => {
                        switch (key) {
                            case 'topic': {
                                values.push((value as any).toString());
                                break;
                            }
                        }
                        return values;
                    },
                    []
                );
                resultFound = false;
                for (let i = 0; i < rowValues.length; i++) {
                    const rowValue = rowValues[i];
                    if (rowValue.includes(searchValue)) {
                        resultFound = true;
                        break;
                    }
                }
            }
            return resultFound;
        }
    );
    return filteredRows;
};
const onChangeUserId = (userId: string) => {
    currentUserId.value = userId;
    store.setCurrentUserId(userId);
    errorOccurred.value = false;
    $q.loading.show();
    getRecordings(userId)
        .then((recordings: IZoomMeetingRecordingsResponseItem[]) => {
            data.value = recordings;
            store.setRecordings(recordings);
        })
        .catch(() => {
            errorOccurred.value = true;
            showErrorMessage();
        })
        .finally(() => {
            $q.loading.hide();
        });
};
const onRefresh = () => {
    const userId = currentUserId.value;
    if (userId) {
        errorOccurred.value = false;
        $q.loading.show();
        getRecordings(userId)
            .then((recordings) => {
                data.value = recordings;
                store.setRecordings(recordings);
            })
            .catch(() => {
                errorOccurred.value = true;
                showErrorMessage();
            })
            .finally(() => {
                $q.loading.hide();
            });
    }
};
const onSearch = () => {
    store.setSearch(search.value);
};
const onCopy = () => {
    if (playUrl.value) {
        const message = 'URL copied to the clipboard!';
        copyInput(playUrl.value, message).then(() => {
            playUrl.value = '';
        });
    }
};
const onOpen = () => {
    if (playUrl.value) {
        const redirectWindow = window.open(playUrl.value, '_blank');
        redirectWindow?.location;
    }
};
const showErrorMessage = () => {
    if (errorOccurred.value) {
        toast('An error occurred!', ELogLevel.ERROR);
    }
};
</script>
