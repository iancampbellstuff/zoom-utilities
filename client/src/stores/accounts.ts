// externals
import { defineStore } from 'pinia';
// types
import { IZoomAccountDataResponseItem } from '../../../common/src/types';

export interface IAccountsState {
    accountData: IZoomAccountDataResponseItem[];
    search: string;
}

export const useAccountsStore = defineStore({
    id: 'accounts',
    state: (): IAccountsState => ({
        accountData: [],
        search: ''
    }),
    actions: {
        resetAll() {
            this.resetAccountData();
            this.resetSearch();
        },
        resetAccountData() {
            this.accountData = [];
        },
        resetSearch() {
            this.search = '';
        },
        setAccountData(accountData: IZoomAccountDataResponseItem[]) {
            this.accountData = accountData;
        },
        setSearch(search: string) {
            this.search = search;
        }
    },
    getters: {
        getAccountData(state) {
            return state.accountData;
        },
        getSearch(state) {
            return state.search;
        }
    }
});
