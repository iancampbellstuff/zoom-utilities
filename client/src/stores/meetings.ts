import { defineStore } from 'pinia';
import { IZoomMeeting } from '../../../common/src';

export interface IMeetingsState {
    currentUserId: string;
    meetings: IZoomMeeting[];
    search: string;
    userIds: string[];
}

export const useMeetingsStore = defineStore({
    id: 'meetings',
    state: (): IMeetingsState => ({
        currentUserId: '',
        meetings: [],
        search: '',
        userIds: []
    }),
    actions: {
        resetAll() {
            this.resetCurrentUserId();
            this.resetMeetings();
            this.resetSearch();
            this.resetUserIds();
        },
        resetCurrentUserId() {
            this.currentUserId = '';
        },
        resetMeetings() {
            this.meetings = [];
        },
        resetSearch() {
            this.search = '';
        },
        resetUserIds() {
            this.userIds = [];
        },
        setCurrentUserId(currentUserId: string) {
            this.currentUserId = currentUserId;
        },
        setMeetings(meetings: IZoomMeeting[]) {
            this.meetings = meetings;
        },
        setSearch(search: string) {
            this.search = search;
        },
        setUserIds(userIds: string[]) {
            this.userIds = userIds;
        }
    },
    getters: {
        getCurrentUserId(state) {
            return state.currentUserId;
        },
        getMeetings(state) {
            return state.meetings;
        },
        getSearch(state) {
            return state.search;
        },
        getUserIds(state) {
            return state.userIds;
        }
    }
});
