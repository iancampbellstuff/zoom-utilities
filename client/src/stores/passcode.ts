import { defineStore } from 'pinia';

export interface IPasscodeState {
    firstDigit?: number;
    length?: number;
    passcode?: number;
}

export const usePasscodeStore = defineStore({
    id: 'passcode',
    state: (): IPasscodeState => ({}),
    actions: {
        resetAll() {
            this.resetFirstDigit();
            this.resetLength();
            this.resetPasscode();
        },
        resetFirstDigit() {
            delete this.firstDigit;
        },
        resetLength() {
            delete this.length;
        },
        resetPasscode() {
            delete this.passcode;
        },
        setFirstDigit(firstDigit?: number) {
            if (firstDigit) {
                this.firstDigit = firstDigit;
            }
        },
        setLength(length?: number) {
            if (length) {
                this.length = length;
            }
        },
        setPasscode(passcode?: number) {
            if (passcode) {
                this.passcode = passcode;
            }
        }
    },
    getters: {
        getFirstDigit(state) {
            return state.firstDigit;
        },
        getLength(state) {
            return state.length;
        },
        getPasscode(state) {
            return state.passcode;
        }
    }
});
