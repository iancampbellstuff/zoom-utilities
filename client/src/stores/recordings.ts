// externals
import { defineStore } from 'pinia';
// types
import { IZoomMeetingRecordingsResponseItem } from '../../../common/src/types';

export interface IRecordingsState {
    currentUserId: string;
    recordings: IZoomMeetingRecordingsResponseItem[];
    search: string;
    userIds: string[];
}

export const useRecordingsStore = defineStore({
    id: 'recordings',
    state: (): IRecordingsState => ({
        currentUserId: '',
        recordings: [],
        search: '',
        userIds: []
    }),
    actions: {
        resetAll() {
            this.resetCurrentUserId();
            this.resetRecordings();
            this.resetSearch();
            this.resetUserIds();
        },
        resetCurrentUserId() {
            this.currentUserId = '';
        },
        resetRecordings() {
            this.recordings = [];
        },
        resetSearch() {
            this.search = '';
        },
        resetUserIds() {
            this.userIds = [];
        },
        setCurrentUserId(currentUserId: string) {
            this.currentUserId = currentUserId;
        },
        setRecordings(recordings: IZoomMeetingRecordingsResponseItem[]) {
            this.recordings = recordings;
        },
        setSearch(search: string) {
            this.search = search;
        },
        setUserIds(userIds: string[]) {
            this.userIds = userIds;
        }
    },
    getters: {
        getCurrentUserId(state) {
            return state.currentUserId;
        },
        getRecordings(state) {
            return state.recordings;
        },
        getSearch(state) {
            return state.search;
        },
        getUserIds(state) {
            return state.userIds;
        }
    }
});
