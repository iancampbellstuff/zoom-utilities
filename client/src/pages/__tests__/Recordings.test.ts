import { mount, VueWrapper } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import Quasar from 'quasar';
import { useRecordingsStore } from '../../stores';
import Recordings from '../Recordings.vue';

jest.mock('quasar', () => {
    const Notify = {
        create: jest.fn()
    };
    const useQuasar = () => ({
        loading: {
            hide: jest.fn(),
            show: jest.fn()
        }
    });
    return {
        ...jest.requireActual('quasar'),
        Notify,
        useQuasar
    };
});

describe('Recordings', () => {
    let store;
    let wrapper: VueWrapper<any>;
    beforeEach(() => {
        store = useRecordingsStore(createTestingPinia());
        wrapper = mount(Recordings, {
            global: {
                plugins: [Quasar]
            },
            store,
            stubs: [
                'q-btn',
                'q-card-actions',
                'q-card-section',
                'q-card',
                'q-dialog',
                'q-form',
                'q-icon',
                'q-input',
                'q-page',
                'q-select',
                'q-table',
                'q-td',
                'q-toolbar-title',
                'q-tr'
            ]
        });
    });
    test('is a Vue instance', () => {
        expect(wrapper.vm).toBeTruthy();
    });
    it('should render correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });
});
