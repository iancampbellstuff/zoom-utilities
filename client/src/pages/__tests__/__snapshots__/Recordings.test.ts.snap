// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recordings should render correctly 1`] = `
<q-page
  class="row items-start justify-evenly"
  id="recordings"
>
  <q-form
    class="q-gutter-lg"
    id="form"
  >
    <q-toolbar-title
      class="text-center"
    >
      Recordings
    </q-toolbar-title>
    <q-table
      columns="[object Object],[object Object],[object Object],[object Object]"
      filter=""
      filter-method="function (rows, searchValue) {
      /* istanbul ignore next */
      cov_2kfo8v29i9().f[9]++;
      var filteredRows =
      /* istanbul ignore next */
      (cov_2kfo8v29i9().s[41]++, rows.filter(function (row) {
        /* istanbul ignore next */
        cov_2kfo8v29i9().f[10]++;
        var resultFound =
        /* istanbul ignore next */
        (cov_2kfo8v29i9().s[42]++, true);
        /* istanbul ignore next */
        cov_2kfo8v29i9().s[43]++;
        if (searchValue !== '') {
          /* istanbul ignore next */
          cov_2kfo8v29i9().b[3][0]++;
          var rowValues =
          /* istanbul ignore next */
          (cov_2kfo8v29i9().s[44]++, Object.entries(row).reduce(function (values, _a) {
            /* istanbul ignore next */
            cov_2kfo8v29i9().f[11]++;
            var key =
              /* istanbul ignore next */
              (cov_2kfo8v29i9().s[45]++, _a[0]),
              value =
              /* istanbul ignore next */
              (cov_2kfo8v29i9().s[46]++, _a[1]);
            /* istanbul ignore next */
            cov_2kfo8v29i9().s[47]++;
            switch (key) {
              case 'topic':
                /* istanbul ignore next */
                cov_2kfo8v29i9().b[4][0]++;
                {
                  /* istanbul ignore next */
                  cov_2kfo8v29i9().s[48]++;
                  values.push(value.toString());
                  /* istanbul ignore next */
                  cov_2kfo8v29i9().s[49]++;
                  break;
                }
            }
            /* istanbul ignore next */
            cov_2kfo8v29i9().s[50]++;
            return values;
          }, []));
          /* istanbul ignore next */
          cov_2kfo8v29i9().s[51]++;
          resultFound = false;
          /* istanbul ignore next */
          cov_2kfo8v29i9().s[52]++;
          for (var i =
          /* istanbul ignore next */
          (cov_2kfo8v29i9().s[53]++, 0); i < rowValues.length; i++) {
            var rowValue =
            /* istanbul ignore next */
            (cov_2kfo8v29i9().s[54]++, rowValues[i]);
            /* istanbul ignore next */
            cov_2kfo8v29i9().s[55]++;
            if (rowValue.includes(searchValue)) {
              /* istanbul ignore next */
              cov_2kfo8v29i9().b[5][0]++;
              cov_2kfo8v29i9().s[56]++;
              resultFound = true;
              /* istanbul ignore next */
              cov_2kfo8v29i9().s[57]++;
              break;
            } else
            /* istanbul ignore next */
            {
              cov_2kfo8v29i9().b[5][1]++;
            }
          }
        } else
        /* istanbul ignore next */
        {
          cov_2kfo8v29i9().b[3][1]++;
        }
        cov_2kfo8v29i9().s[58]++;
        return resultFound;
      }));
      /* istanbul ignore next */
      cov_2kfo8v29i9().s[59]++;
      return filteredRows;
    }"
      outlined=""
      pagination="[object Object]"
      row-key="id"
      rows=""
      standout=""
      title="Recordings"
    />
  </q-form>
</q-page>
`;
